packages/edit-site/src/components/block-editor/use-site-editor-settings.js

```diff
@@ -92,13 +93,18 @@ function useArchiveLabel( templateSlug ) {
 
 function useGoBack() {
 	const location = useLocation();
+	const previousLocation = usePrevious( location );
 	const history = useHistory();
 	const goBack = useMemo( () => {
 		const isFocusMode =
 			location.params.focusMode ||
-			FOCUSABLE_ENTITIES.includes( location.params.postType );
-		return isFocusMode ? () => history.back() : undefined;
-	}, [ location.params.focusMode, location.params.postType, history ] );
+			( location.params.postId &&
+				FOCUSABLE_ENTITIES.includes( location.params.postType ) );
+		const didComeFromEditorCanvas =
+			previousLocation?.params.canvas === 'edit';
+		const showBackButton = isFocusMode && didComeFromEditorCanvas;
+		return showBackButton ? () => history.back() : undefined;
+	}, [ location ] );
```

Comment:
This PR's testing nicely for me! Just noticed that the linter is complaining about missing items in the dependency array for this one:

<img width="590" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/fa144490-27b7-4dcd-83f6-427de4133403">

Is it okay to include those, too?

Reply:
Hm. When I add `previousLocation` everything stops working as it seems to update too often. Maybe `previousLocation` isn't a stable reference....

How much of a "rule" is this lint rule? 😀

Reply:
I see lots of `// eslint-disable-next-line react-hooks/exhaustive-deps` in the code base. I think it's one of those rules that we should attempt to honour when we can, but if we can't, then it's good to add in that `disable` line with a line or two with a `Ignore reason` or `Disable reason` with a quick explanation 🙂

Reply:
OK I figured it out. `usePrevious` had a flawed implementation in that it returns what `value` was the previous time that the component rendered. This is different to what we want, and what I think most consumers want, which is that `usePrevious` should return what `value` previously was.

If the component updates for some reason unrelated to `value` changing (e.g. its state changed) then we'll get the current `value` instead of the previous `value`.

I found this article which explains the problem and solution really well: 

https://www.developerway.com/posts/implementing-advanced-use-previous-hook

I also saw that the implementation that we based our version of `usePrevious` on has been changed to be similar to the approach suggested in the article above:

https://github.com/uidotdev/usehooks/blob/main/index.js#L1017

So, I updated our implementation of `usePrevious` to follow. I also added a unit test. Now, when I set `deps` to be exhaustive, the back button updates properly.

Reply:
Great catch — I think this is probably the kind of flaw that the folks who introduced that linting rule were hoping we'd uncover!

It looks like the unit tests are failing now, in particular some ToolsPanel tests, which use `usePrevious` internally. If that winds up being complex to solve, since `usePrevious` / the compose package is a bit lower level than the proposed change in this PR, would it be worth splitting out the idea into two parts — 1) add in the `eslint-disable-next-line` for this PR to get things working for now, and 2) open a separate PR for `usePrevious` where we can debug the `ToolsPanel` test, and see if there were components that expected the previous implementation?

Reply:
Yeah, bugger, let's do that.